File: app/src/main/java/org/d1scw0rld/bookbag/fields/FieldAutoCompleteTextView.java
package org.d1scw0rld.bookbag.fields;

import java.util.ArrayList;

import org.d1scw0rld.bookbag.FilteredArrayAdapter;
import org.d1scw0rld.bookbag.R;
import org.d1scw0rld.bookbag.dto.Property;
import org.d1scw0rld.bookbag.fields.AutoCompleteTextViewX.Callback;
import android.content.Context;
import android.content.res.TypedArray;
import android.text.Editable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;

public class FieldAutoCompleteTextView extends LinearLayout implements Field
{
   
   private Title                 title;
   private AutoCompleteTextViewX autoCompleteTextViewX;

   public FieldAutoCompleteTextView(Context context)
   {
      super(context);
      
      init(context);
   }

   public FieldAutoCompleteTextView(Context context, AttributeSet attrs)
   {
      super(context, attrs);
      
      init(context);
      
      TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.FieldAutoCompleteTextView, 0, 0);

      String titleText = a.getString(R.styleable.FieldAutoCompleteTextView_title);
      int titleValueColor = a.getColor(R.styleable.FieldAutoCompleteTextView_titleColor, 0);
      int titleTextSize = a.getDimensionPixelOffset(R.styleable.FieldAutoCompleteTextView_titleTextSize, 0);
      int titleLineSize = a.getDimensionPixelOffset(R.styleable.FieldAutoCompleteTextView_titleLineSize, 0);
      String text = a.getString(R.styleable.FieldAutoCompleteTextView_android_text);
      String hint = a.getString(R.styleable.FieldAutoCompleteTextView_android_hint);
      
      a.recycle();

      setOrientation(LinearLayout.VERTICAL);
      setGravity(Gravity.CENTER_VERTICAL);

      title.setText(titleText);
      title.setColor(titleValueColor);
      title.setTextSize(titleTextSize);
      title.setLineSize(titleLineSize);
      
      autoCompleteTextViewX.setText(text);
      autoCompleteTextViewX.setHint(hint);
   }

   public void init(Context context)
   {
      LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
      inflater.inflate(R.layout.field_auto_complete_text_view, this, true);

      title = this.findViewById(R.id.title);
      autoCompleteTextViewX = this.findViewById(R.id.autoCompleteTextView);
      autoCompleteTextViewX.setThreshold(1);
   }

   public void setTitle(String title)
   {
      this.title.setText(title);
   }

   public void setTitle(int resid)
   {
      title.setText(resid);
   }

   @Override
   public String getTitle()
   {
      return title.getTitle();
   }

   public void setTitleColor(int valueColor)
   {
      title.setColor(valueColor);
   }
   
   public void setTitleTextSize(int textSize)
   {
      title.setTextSize(textSize);
   }
   
   public void setLineSize(int lineSize)
   {
      title.setTextSize(lineSize);
   }
   
   public void setText(String text)
   {
      autoCompleteTextViewX.setText(text);
   }

   public void setText(int resid)
   {
      autoCompleteTextViewX.setText(resid);
   }
   
   public Editable getText()
   {
      return autoCompleteTextViewX.getText();
   }
   
   public void setHint(String hint)
   {
      autoCompleteTextViewX.setHint(hint);
   }

   public void setHint(int resid)
   {
      autoCompleteTextViewX.setHint(resid);
   }
   
   public void setThreshold(int threshold)
   {
      autoCompleteTextViewX.setThreshold(threshold);
   }
   
   public void setAdapter(ArrayAdapter<?> adapter)
   {
      autoCompleteTextViewX.setAdapter(adapter);
   }
   
   public void setUpdateListener(AutoCompleteTextViewX.OnUpdateListener onUpdateListener)
   {
      autoCompleteTextViewX.setOnUpdateListener(onUpdateListener);
   }

   public void setCallback(Callback callback)
   {
      autoCompleteTextViewX.setCallback(callback);
   }   
   
   public void setOnItemSelectedListener(OnItemSelectedListener l)
   {
      autoCompleteTextViewX.setOnItemSelectedListener(l);
   }

   public void setOnItemClickListener(OnItemClickListener l)
   {
      autoCompleteTextViewX.setOnItemClickListener(l);
   }
}

File: app/src/main/java/org/d1scw0rld/bookbag/fields/FieldAutoCompleteTextView.kt
package org.d1scw0rld.bookbag.fields

import android.content.Context
import android.text.Editable
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import org.d1scw0rld.bookbag.R

class FieldAutoCompleteTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : LinearLayout(context, attrs, defStyleAttr), Field {

    private lateinit var title: Title
    private lateinit var autoCompleteTextViewX: AutoCompleteTextViewX

    init {
        initialize(context)

        orientation = VERTICAL
        gravity = Gravity.CENTER_VERTICAL

        attrs?.let {
            val typedArray = context.obtainStyledAttributes(it, R.styleable.FieldAutoCompleteTextView, 0, 0)

            val titleText = typedArray.getString(R.styleable.FieldAutoCompleteTextView_title)
            val titleValueColor = typedArray.getColor(R.styleable.FieldAutoCompleteTextView_titleColor, 0)
            val titleTextSize = typedArray.getDimensionPixelOffset(R.styleable.FieldAutoCompleteTextView_titleTextSize, 0)
            val titleLineSize = typedArray.getDimensionPixelOffset(R.styleable.FieldAutoCompleteTextView_titleLineSize, 0)
            val text = typedArray.getString(R.styleable.FieldAutoCompleteTextView_android_text)
            val hint = typedArray.getString(R.styleable.FieldAutoCompleteTextView_android_hint)

            typedArray.recycle()

            titleText?.let { t -> this.title.setText(t) }
            this.title.setColor(titleValueColor)
            this.title.setTextSize(titleTextSize)
            this.title.setLineSize(titleLineSize)

            autoCompleteTextViewX.setText(text)
            autoCompleteTextViewX.hint = hint
        }
    }

    private fun initialize(context: Context) {
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.field_auto_complete_text_view, this, true)

        title = findViewById(R.id.title)
        autoCompleteTextViewX = findViewById(R.id.autoCompleteTextView)
        autoCompleteTextViewX.threshold = 1
    }

    override fun setTitle(text: String) {
        title.setText(text)
    }

    override fun setTitle(resid: Int) {
        title.setText(resid)
    }

    override fun getTitle(): String {
        return title.getTitle()
    }

    override fun setTitleColor(valueColor: Int) {
        title.setColor(valueColor)
    }

    override fun setTitleTextSize(textSize: Int) {
        title.setTextSize(textSize)
    }

    fun setLineSize(lineSize: Int) {
        title.setLineSize(lineSize)
    }

    fun setText(text: String?) {
        autoCompleteTextViewX.setText(text)
    }

    fun setText(resourceId: Int) {
        autoCompleteTextViewX.setText(resourceId)
    }

    fun getText(): Editable? {
        return autoCompleteTextViewX.text
    }

    fun setHint(hint: String?) {
        autoCompleteTextViewX.hint = hint
    }

    fun setHint(resourceId: Int) {
        autoCompleteTextViewX.setHint(resourceId)
    }

    fun setThreshold(threshold: Int) {
        autoCompleteTextViewX.threshold = threshold
    }

    fun setAdapter(adapter: ArrayAdapter<*>?) {
        autoCompleteTextViewX.setAdapter(adapter)
    }

    fun setUpdateListener(onUpdateListener: AutoCompleteTextViewX.OnUpdateListener?) {
        autoCompleteTextViewX.setOnUpdateListener(onUpdateListener)
    }

    fun setCallback(callback: AutoCompleteTextViewX.Callback?) {
        autoCompleteTextViewX.setCallback(callback)
    }

    fun setOnItemSelectedListener(listener: OnItemSelectedListener?) {
        autoCompleteTextViewX.onItemSelectedListener = listener
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        autoCompleteTextViewX.onItemClickListener = listener
    }
}
